docs/proposals/20201123-tracing-ingestion-api-and-opentelemetry-collector.md

```diff
@@ -0,0 +1,91 @@
+# Traces ingestion API and OpenTelemetry collector
+
+* **Owners:**:
+    * `@pavolloffay`
+
+* **Related Tickets:**
+    * [observatorium/issues/442](https://github.com/observatorium/observatorium/issues/442)
+
+## Table of Contents
+
+- [TLDR](#tldr)
+- [Why](#why)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [How](#how)
+  * [Expose tracing ingestion API in the Observatorium API service](#expose-tracing-ingestion-api-in-the-observatorium-api-service)
+    + [Multitenancy](#multitenancy)
+  * [Add deployments manifests for OpenTelemetry collector](#add-deployments-manifests-for-opentelemetry-collector)
+- [Action Plan](#action-plan)
+- [References](#references)
+
+<small>
+
+<i>
+
+<a href="http://ecotrust-canada.github.io/markdown-toc/">
+Table of contents generated with markdown-toc
+</a>
+
+</i>
+
+</small>
+
+## TLDR
+
+This document proposes to expose tracing ingestion API  backed by the OpenTelemetry collector. This is a first step to support ingestion and storage for tracing data. The OpenTelemetry collector gives flexibility to forward tracing data to variety of tracing platforms, both on-prem or SaaS - Jaeger, Zipkin, Tempo, SigNoz or DataDog, Dynatrace etc.
```

Comment:
Hmm, while it sounds great, I don't see why the backend needs such flexibility. We own this path, we can add support of anything to the solution we would use. 

Besides that, sounds to me like our main potential solutions are Jaeger and Tempo and both support just the OTLP tracing protocol, right? 🤔 

PS: And why do we need DataDog? In such case, there is no need for Observatorium, no?

If anyone wants flexibility they can instal Otel collector before Observatorium. IMO we should make as simple an approach as possible. Complexity with generic flexibility has a major cost.

> Could you please explain "such a heavy dependency". OTEL collector is quite a small component that just receives, transforms and forwards data. I will add some reasoning to the text why we should add the collector.

Collector main distro (I assume we want to use that) itself has ~35 dependencies. Repo has 200 issue despite impressive amount of activity and maintainers - that is NOT a lightweight dependency - and easy to maintain. And if we want to put this as a sidecar that is actually a massive complexity addition (another process to maintain, monitor, understand, debug if it's performance hit, etc). Let's put it in different perspective. Adding Otel collector as sidecar to Observatorium API proxy is like adding Prometheus next to Observatorium API. In terms of complexity and size of those projects, our API proxy is tiny. I don't think Otel collector is a free lunch we can treat as black box here.

In order to achieve what we want (tenancy, auth, rate limit) I believe we could get this working correctly for one protocol first. I am afraid making the number of permutations exponential of different outputs (or inputs) will only drop our focus.

At the end, it's really ~100 lines of code to add tag / check stuff per HTTP/gRPC message. Instead of complex collector code, I think that's a win.

Reply:
I think if we could take a step back and summarize what we would want to achieve and what are the options. I think we can agree on wanting the API to accept spans in OTLP protocol (whether gRPC or HTTP) as our initial position.

The question is, after ingesting the traces, what are our options to export the spans. So far as I understand the two sides:
1. We can send the traces to Otel Collector, which will forward them to appropriate tracing backend
 - the downside: we need to introduce an extra component (collector) between API and tracing backend = extra dependency, extra overhead etc.
 
2. We send traces to a tracing backend which is directly capable of accepting OTLP and remove the collector
- the downside: this leaves out Jaeger, which does not seem to support OTLP currently; What is the alternative here? @bwplotka you suggested Tempo? How big of a dependency is that in comparison to collector+Jaeger? What are it's other (dis)advantages?

Am I understanding this correctly? I believe we should capture these points in the chosen solution and in the alternatives.

Reply:
agree with @matej-g , I think would help systematically listing pros/cons for each option

> At the end, it's really ~100 lines of code to add tag / check stuff per HTTP/gRPC message. Instead of complex collector code, I think that's a win.

@bwplotka could you expand more on this? you mean we could write the necessary/custom code to support tracing for the OTLP protocol instead of adding the collector as a dependency?

Reply:
Why doesn't OT support multi-tenancy? The ot-collector is intended as a vendor-agnostic implementation on how to receive, process and export telemetry data (including metrics/logs/tracing). The primary disjunct is that Observatorium is positioned as an _opinionated_, _multi-tenanted_ solution to the same problem. OT doesn't currently support soft-multi tenancy. So it seems contradictory to incur the complexity/overhead of maintaining stack agnostic components when the intention is to be opinionated. 

I think it's worth enumerating the core use-case here for the initial tracing implementation. Is it critical to have Jaeger support? Is the OT-collector the only interface we might have with Jaeger?